File: src/models/balance.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::clock::Clock;

use super::Asset;

/// A single asset's balance without timestamp (belongs to a snapshot).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AssetBalance {
    pub asset: Asset,
    /// Amount as string to avoid floating point precision issues
    pub amount: String,
    /// Optional total cost basis for this holding, denominated in the
    /// portfolio/reporting currency used for gain estimates.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cost_basis: Option<String>,
}

impl AssetBalance {
    pub fn new(asset: Asset, amount: impl Into<String>) -> Self {
        Self {
            asset,
            amount: amount.into(),
            cost_basis: None,
        }
    }

    pub fn with_cost_basis(mut self, cost_basis: impl Into<String>) -> Self {
        self.cost_basis = Some(cost_basis.into());
        self
    }
}

/// A point-in-time snapshot of ALL holdings in an account.
/// One line in the JSONL file = one complete account state.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BalanceSnapshot {
    pub timestamp: DateTime<Utc>,
    pub balances: Vec<AssetBalance>,
}

impl BalanceSnapshot {
    pub fn new(timestamp: DateTime<Utc>, balances: Vec<AssetBalance>) -> Self {
        Self {
            timestamp,
            balances,
        }
    }

    pub fn now(balances: Vec<AssetBalance>) -> Self {
        Self::new(Utc::now(), balances)
    }

    pub fn now_with(clock: &dyn Clock, balances: Vec<AssetBalance>) -> Self {
        Self::new(clock.now(), balances)
    }
}

File: ts/src/models/balance.ts
import { type AssetType } from './asset.js';
import { Clock, SystemClock } from '../clock.js';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface AssetBalanceType {
  readonly asset: AssetType;
  readonly amount: string;
  readonly cost_basis?: string;
}

export interface BalanceSnapshotType {
  readonly timestamp: Date;
  readonly timestamp_raw?: string;
  readonly balances: AssetBalanceType[];
}

// ---------------------------------------------------------------------------
// JSON types
// ---------------------------------------------------------------------------

export interface AssetBalanceJSON {
  asset: AssetType;
  amount: string;
  cost_basis?: string;
}

export interface BalanceSnapshotJSON {
  timestamp: string;
  balances: AssetBalanceJSON[];
}

// ---------------------------------------------------------------------------
// AssetBalance namespace
// ---------------------------------------------------------------------------

export const AssetBalance = {
  /**
   * Create an asset balance.
   */
  new(asset: AssetType, amount: string, costBasis?: string): AssetBalanceType {
    return costBasis === undefined ? { asset, amount } : { asset, amount, cost_basis: costBasis };
  },

  /**
   * Serialize to JSON.
   */
  toJSON(balance: AssetBalanceType): AssetBalanceJSON {
    const out: AssetBalanceJSON = {
      asset: balance.asset,
      amount: balance.amount,
    };
    if (balance.cost_basis !== undefined) {
      out.cost_basis = balance.cost_basis;
    }
    return out;
  },

  /**
   * Deserialize from JSON.
   */
  fromJSON(json: AssetBalanceJSON): AssetBalanceType {
    const out: AssetBalanceType = {
      asset: json.asset,
      amount: json.amount,
    };
    if (json.cost_basis !== undefined) {
      return { ...out, cost_basis: json.cost_basis };
    }
    return out;
  },
} as const;

// ---------------------------------------------------------------------------
// BalanceSnapshot namespace
// ---------------------------------------------------------------------------

export const BalanceSnapshot = {
  /**
   * Create a balance snapshot with a given timestamp and balances.
   */
  new(timestamp: Date, balances: AssetBalanceType[]): BalanceSnapshotType {
    return { timestamp, balances: [...balances] };
  },

  /**
   * Create a balance snapshot at the current time.
   */
  now(balances: AssetBalanceType[]): BalanceSnapshotType {
    return BalanceSnapshot.nowWith(new SystemClock(), balances);
  },

  /**
   * Create a balance snapshot using an injected clock.
   */
  nowWith(clock: Clock, balances: AssetBalanceType[]): BalanceSnapshotType {
    return BalanceSnapshot.new(clock.now(), balances);
  },

  /**
   * Serialize to JSON.
   */
  toJSON(snapshot: BalanceSnapshotType): BalanceSnapshotJSON {
    return {
      timestamp: snapshot.timestamp_raw ?? snapshot.timestamp.toISOString(),
      balances: snapshot.balances.map(AssetBalance.toJSON),
    };
  },

  /**
   * Deserialize from JSON.
   */
  fromJSON(json: BalanceSnapshotJSON): BalanceSnapshotType {
    return {
      timestamp: new Date(json.timestamp),
      timestamp_raw: json.timestamp,
      balances: json.balances.map(AssetBalance.fromJSON),
    };
  },
} as const;
